src/System.Private.CoreLib/shared/System/Collections/Generic/Dictionary.cs

```diff
@@ -318,102 +326,132 @@ public virtual void GetObjectData(SerializationInfo info, StreamingContext conte
             }
         }
 
-        private int FindEntry(TKey key)
+        private unsafe ref TValue FindValue(TKey key)
         {
             if (key == null)
             {
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.key);
             }
 
-            int i = -1;
             int[]? buckets = _buckets;
-            Entry[]? entries = _entries;
-            int collisionCount = 0;
+            ref Entry entry = ref Unsafe.AsRef<Entry>(null);
             if (buckets != null)
             {
-                Debug.Assert(entries != null, "expected entries to be != null");
+                Debug.Assert(_entries != null, "expected entries to be != null");
                 IEqualityComparer<TKey>? comparer = _comparer;
                 if (comparer == null)
                 {
                     uint hashCode = (uint)key.GetHashCode();
-                    // Value in _buckets is 1-based
-                    i = buckets[hashCode % (uint)buckets.Length] - 1;
+                    int i = buckets[hashCode % (uint)buckets.Length];
                     if (default(TKey)! != null) // TODO-NULLABLE: default(T) == null warning (https://github.com/dotnet/roslyn/issues/34757)
                     {
                         // ValueType: Devirtualize with EqualityComparer<TValue>.Default intrinsic
-                        while (true)
+                        Entry[]? entries = _entries;
+                        int collisionCount = 0;
```

Comment:
You had previously tried an experiment where you had something like `int allowedCollisionsRemaining = entries.Length;`, then decremented this number at the end of the loop and checked that the number was non-zero. Did this not pan out in practice? In theory this would allow the `add` and `jnz` instructions at the end of the loop to fuse. See https://www.agner.org/optimize/microarchitecture.pdf (pp. 125 - 126).

Reply:
Always outputs `dec` and `test`; test fuses, but it could also be dropped and `dec` will fuse, but the Jit doesn't do that https://github.com/dotnet/coreclr/issues/7566 😢 
```
dec      r13d
test     r13d, r13d
jl       SHORT G_M59263_IG15
```
I've moved the addition to before the jump rather than after so it could still fuse; except it still inserts a `test` so doesn't. 

I did spike up a Jit change for it https://github.com/dotnet/coreclr/compare/master...benaadams:peephole-opt-incdec?expand=1 but there is weirdness around the Overflow flag with inc/dec; so left it alone.

Reply:
A `jl` instruction makes it seem like this was a comparison against `< 0` rather than `== 0`?

Reply:
I tried against 0 and it didn't work :( https://github.com/dotnet/coreclr/issues/7566#issuecomment-541439929

Reply:
Blergh. :(

Reply:
Note that in the "good" case with no collisions there is no perf benefit in doing this change.  Long collision chains are not expected and may indicate a bad hash function.

Reply:
Or a full Dictionary because someone called `.TrimExcess()`; the "increase my collisions" method 😉